### DISCUSSION TITLE
2FA key in twofactor.data field are not encrypted

### DISCUSSION THREAD

**Discussion by teraposta** [dani-garcia/vaultwarden#5255] [General]
I noticed that 2FA keys are stored unencrypted in the twofactor table, which could pose a security risk if the database is compromised.

I was wondering if the same design choice exists in the upstream Bitwarden implementation.

Thanks

**Comment by BlackDex** [comment]
This is also done upstream: 
![image](https://github.com/user-attachments/assets/9935a842-f327-402f-b9ba-1e0046b62887)

You need to be able to use the original key to verify if the TOTP is correct.
The only way to encrypt something like this is to use a secret key which needs to be linked to the database and should always stay the same no matter what.

The main point is that, if someone could gain access to your database, they probably also could extract that key to decrypt the data.
There are some techniques which could make it a bit harder, using a different table to store some other data. But in the end, it still is something which can be decrypted and not hashed for example, like a password.

**Reply by teraposta** [reply]
Thank you for your insightful response. I understand that the requirement for the encryption key to remain consistent unless deliberately changed by the user limits the effectiveness of implementing meaningful encryption for the stored keys. While additional measures (e.g., splitting data across tables) could add a layer of complexity, I recognize that such methods would still not fully eliminate the ability to decrypt the data if the database itself is compromised.
